Sources/Sandbox.Game/Game/Entities/Blocks/ToolbarItem.cs

```diff
@@ -0,0 +1,80 @@
+﻿using System.Collections.Generic;
+using ProtoBuf;
+using Sandbox.Common.ObjectBuilders;
+using Sandbox.Game.Entities.Cube;
+using Sandbox.Game.Screens.Helpers;
+
+namespace Sandbox.Game.Entities.Blocks
+{
+    [ProtoContract]
+    internal struct ToolbarItem
+    {
+        public static ToolbarItem FromItem(MyToolbarItem item)
+        {
+            var tItem = new ToolbarItem();
+            tItem.EntityID = 0;
+            var terminalItem = item as MyToolbarItemTerminalBlock;
+            if (terminalItem != null)
+            {
+                var block = item.GetObjectBuilder() as MyObjectBuilder_ToolbarItemTerminalBlock;
+                tItem.EntityID = block.BlockEntityId;
+                tItem.Action = block.Action;
+                tItem.Parameters = block.Parameters;
+            }
+            else if (item is MyToolbarItemTerminalGroup)
+            {
+                var block = item.GetObjectBuilder() as MyObjectBuilder_ToolbarItemTerminalGroup;
+                tItem.EntityID = block.BlockEntityId;
+                tItem.Action = block.Action;
+                tItem.GroupName = block.GroupName;
+                tItem.Parameters = block.Parameters;
+            }
+            return tItem;
+        }
+
+        public static MyToolbarItem ToItemFor<T>(ToolbarItem msgItem) where T : MyCubeBlock
```

Comment:
Please make it non-generic, it should be enough. Only generic usage is TryGetEntityById, but it still uses only CubeGrid property. Should be enough to TryGetEntityById<MyCubeBlock>.